### PR TITLE
handle long class names for generated java classes

### PR DESCRIPTION
**Pull request by MauricioUyaguari**
handle generating junit test classes for long element names that along with the prefix (Test) and extension (.java) hit the 255 limit. 
For this we will append the hashCode of the element name to the end of the name replacing chars so that it doesn't go past limit.

Addresses any generated class that goes over the limit but this PR was created to fix function tests which generally have longer names since they append parameterss and return type to their name for function overloading.

### CODE REVIEW

File: legend-sdlc-test-generation/src/main/java/org/finos/legend/sdlc/test/junit/AbstractTestClassGenerator.java [hunk 1]
```diff
@@ -58,7 +64,18 @@ private String generateTestClassPackage(PackageableElement protocolElement)
...
+
+    private String generateTestName(String name)
+    {
+        int maxByteLength = NAME_BYTE_MAX - TEST_PREFIX.getBytes(StandardCharsets.UTF_8).length;
+        if (name.getBytes(StandardCharsets.UTF_8).length > maxByteLength)
+        {
+            String hashCode = String.valueOf(Math.abs(name.hashCode()));
+            return name.substring(0, maxByteLength - hashCode.getBytes(StandardCharsets.UTF_8).length) + hashCode;
+        }
+        return name;
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-test-generation/src/main/java/org/finos/legend/sdlc/test/junit/AbstractTestClassGenerator.java:78]: I think something like this, which will make the renaming a little more uniform, is better.  This also takes care of a problem with handling of unicode characters that require multiple bytes to encode.
```suggestion
        int overage = name.getBytes(StandardCharsets.UTF_8).length - NAME_BYTE_MAX;
        if (overage > 0)
        {
            int newLen = name.length() - overage - 9;
            while (Character.isSurrogate(name.charAt(newLen - 1))
            {
                len--;
            }
            StringBuilder builder = new StringBuilder(newLen + 9);
            builder.append(name, 0, newLen);
            new Formatter(builder).format("_%08x", name.hashCode());
            return builder.toString();
        }
        return name;
```

File: legend-sdlc-test-generation/src/test/java/org/finos/legend/sdlc/test/junit/TestServiceTestGenerator.java [hunk 2]
```diff
@@ -59,6 +59,29 @@ public void testTestService2()
                 "testTestSuites::TestService2");
     }
 
+
+    @Test
+    public void testTestLongServiceName()
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-test-generation/src/test/java/org/finos/legend/sdlc/test/junit/TestServiceTestGenerator.java:64]: You also need tests for names with unicode characters that require multiple bytes to encode, including one that spans the byte limit.

File: legend-sdlc-test-generation/src/test/java/org/finos/legend/sdlc/test/junit/TestJUnitTestGeneratorSerialization.java [hunk 3]
```diff
@@ -61,19 +61,19 @@ public static void cleanUp() throws Exception
...
     public void testWriteTestClassesWithoutRootPackage() throws IOException
     {
-        testWriteTestClasses(null, "generated/java/execution/TestRelationalMapping.java", "generated/java/legend/demo/TestSingleQuoteInResultM2M.java", "generated/java/model/mapping/TestSourceToTargetM2M.java", "generated/java/testTestSuites/TestTestService.java", "generated/java/testTestSuites/TestTestService2.java", "generated/java/testTestSuites/TestServiceStoreMapping.java");
+        testWriteTestClasses(null, "generated/java/execution/TestRelationalMapping.java", "generated/java/execution/TestRelationalMapping.java", "generated/java/legend/demo/TestSingleQuoteInResultM2M.java", "generated/java/model/mapping/TestSourceToTargetM2M.java", "generated/java/testTestSuites/TestTestService.java", "generated/java/testTestSuites/TestTestService2.java", "generated/java/testTestSuites/TestServiceStoreMapping.java", "generated/java/testTestSuites/TestMyServiceIsVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryV953300342.java");
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-test-generation/src/test/java/org/finos/legend/sdlc/test/junit/TestJUnitTestGeneratorSerialization.java:64]: You also need tests for names with unicode characters that require multiple bytes to encode, including one that spans the byte limit.

File: legend-sdlc-test-generation/src/test/java/org/finos/legend/sdlc/test/junit/TestServiceTestGenerator.java [hunk 4]
```diff
@@ -59,6 +59,48 @@ public void testTestService2()
...
+        testGeneration(
+                "other.test.pkg.testTestSuites.TestMyServiceIsVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVe_9dc53473",
+                "generated/java/other/test/pkg/testTestSuites/TestMyServiceIsVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVe_9dc53473.java",
+                "other.test.pkg",
+                "testTestSuites::MyServiceIsVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryâââ");
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-test-generation/src/test/java/org/finos/legend/sdlc/test/junit/TestServiceTestGenerator.java:101]: Can you also add test cases where there are both unicode characters in the middle of the name and also around the cutoff point?

**Review by kevin-m-knight-gs — APPROVED** [review]
